src/advertisement.rs

```diff
@@ -0,0 +1,396 @@
+    self
+      .raw
+      .extend_from_slice(&manufacturer_id.to_le_bytes())
+      .unwrap();
+    self.raw.extend_from_slice(data).unwrap();
+    Ok(self)
+  }
+
+  pub fn push_service_uuids(mut self, uuids: &[UUID]) -> Result<Self, PushError> {
+    let size_of_item = Self::require_equal_size(uuids)?;
+    let ad_type = match uuids.first().ok_or(PushError::UuidInputError)? {
+      UUID::Short(_) => AdType::ServiceData16,
+      UUID::Long(_) => AdType::ServiceData128,
+    };
+
+    self = self.push_start_record(ad_type as _, size_of_item * uuids.len())?;
+    for uuid in uuids {
+      uuid.push_into(&mut self.raw).unwrap();
+    }
+
+    Ok(self)
+  }
+
+  fn require_equal_size(uuids: &[UUID]) -> Result<usize, PushError> {
+    let mut num_bits = None;
+    for uuid in uuids {
+      let size_of = size_of_val(uuid);
+      if num_bits.get_or_insert(size_of) != &size_of {
+        return Err(PushError::UuidInputError);
+      }
+    }
+    num_bits.ok_or(PushError::UuidInputError)
+  }
+
+  pub fn push_service_data(mut self, uuid: &UUID, data: &[u8]) -> Result<Self, PushError> {
+    let (ad_type, size_per_item) = match uuid {
+      UUID::Short(_) => (AdType::ServiceData16, 2),
+      UUID::Long(_) => (AdType::ServiceData128, 16),
+    };
+
+    self = self.push_start_record(ad_type as _, size_per_item + data.len())?;
+    uuid.push_into(&mut self.raw).unwrap();
+    self.raw.extend_from_slice(data).unwrap();
+
+    Ok(self)
+  }
+
+  /// Push the short local name.  It is presumed that this name is either the same or shorter than the
+  /// name assigned to the peripheral itself, and UI treatment from external parties will likely
+  /// reflect this and update the name after connection is established and the "full" name is
+  /// read from the built-in characteristic.  The full name can also be manually pushed using
+  /// [AdType::LongLocalName].
+  pub fn push_local_name(mut self, name: &str) -> Result<Self, PushError> {
+    self = self.push_start_record(AdType::ShortLocalName as _, name.len())?;
+    self.raw.extend_from_slice(name.as_bytes()).unwrap();
+    Ok(self)
+  }
+
+  pub fn push_raw_ad_type(mut self, ad_type: u8, data: &[u8]) -> Result<Self, PushError> {
+    self = self.push_start_record(ad_type, data.len())?;
+    self.raw.extend_from_slice(data).unwrap();
+
+    if ad_type == AdType::Flags as _ {
+      self.has_set_flags = true;
+    }
+
+    Ok(self)
+  }
+
+  fn push_start_record(mut self, ad_type: u8, remaining_size: usize) -> Result<Self, PushError> {
+    self = self.flush_pending_record()?;
+
+    if self.raw.len() + 2 + remaining_size >= N {
+      return Err(PushError::CapacityExceeded);
+    }
+
+    self
+      .raw
+      .push(remaining_size.checked_add(1).unwrap().try_into().unwrap())
+      .unwrap();
+    self.raw.push(ad_type).unwrap();
+    Ok(self)
+  }
+
+  pub fn build(mut self) -> Result<RawAdvertisement<N>, PushError> {
```

Comment:
The main reason was because I wanted to make the API enforce the size constraint at compile time that the protocol defines.  As I understand it slices won't be able to do that?